### PR TITLE
Uploader Extension Point

### PR DESCRIPTION
**Pull request by LawYard**
## Add Extension Point for Uploader

This improvement allow users to create their own Extensions to upload Allure reports into various systems (like Surge) instead of copy report to master only.
Current uploader (Copy report to master) moved to Extension and sets as default.

allure reporter 2 config jenkins 2016-10-28 11-35-02

### CODE REVIEW

File: src/main/java/ru/yandex/qatools/allure/jenkins/AllureBuildAction.java [hunk 1]
```diff
@@ -35,14 +36,23 @@ public String getIconFileName() {
...
     public String getBuildUrl() {
         return build.getUrl();
     }
 
+    public String getReportUrl() {
+        return reportUrl;
+    }
+
     @SuppressWarnings("unused")
     public DirectoryBrowserSupport doDynamic(StaplerRequest req, StaplerResponse rsp)
```
Comment (LawYard) [inline src/main/java/ru/yandex/qatools/allure/jenkins/AllureBuildAction.java]: Not matter. Old warnings

File: src/main/java/ru/yandex/qatools/allure/jenkins/AllureReportUploader.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     *
+     * @param reportDirectory Directory with generated allure report
+     * @param build current build instance
+     * @return url with allure report location
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public abstract String publish(FilePath reportDirectory, AbstractBuild<?, ?> build) throws IOException, InterruptedException, AllureUploadException;
```
Comment (LawYard) [inline src/main/java/ru/yandex/qatools/allure/jenkins/AllureReportUploader.java]: All of this exceptions catched by `AllureReportPublisher`

File: src/main/java/ru/yandex/qatools/allure/jenkins/AllureReportDefaultUploader.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Created by lawyard on 19.09.16.
+ */
+
+
+@Extension
+public class AllureReportDefaultUploader extends AllureReportUploader {
+
+    @DataBoundConstructor
+    public AllureReportDefaultUploader() {}
```
Comment (LawYard) [inline src/main/java/ru/yandex/qatools/allure/jenkins/AllureReportDefaultUploader.java]: Not matter, just warning

File: src/main/java/ru/yandex/qatools/allure/jenkins/AllureReportPublisher.java [hunk 4]
```diff
@@ -200,15 +197,24 @@ private boolean generateReport(List<FilePath> resultsPaths, AbstractBuild<?, ?>
...
             // execute actions for report
-            build.addAction(new AllureBuildAction(build));
+            build.addAction(new AllureBuildAction(build, urlPublished));
         } catch (IOException e) { //NOSONAR
             listener.getLogger().println("Report generation failed");
             e.printStackTrace(listener.getLogger());  //NOSONAR
             return false;
-        } finally {
+        }
+        catch (AllureUploadException e) {
```
Comment (LawYard) [inline src/main/java/ru/yandex/qatools/allure/jenkins/AllureReportPublisher.java]: Exception logged. Look down

File: src/main/java/ru/yandex/qatools/allure/jenkins/AllureReportPublisher.java [hunk 5]
```diff
@@ -200,15 +197,24 @@ private boolean generateReport(List<FilePath> resultsPaths, AbstractBuild<?, ?>
...
+            listener.getLogger().println("Report uploading failed.");
+            e.printStackTrace(listener.getLogger());
```
Comment (LawYard) [inline src/main/java/ru/yandex/qatools/allure/jenkins/AllureReportPublisher.java]: Exception was logged! Sonar sonar....

File: src/main/java/ru/yandex/qatools/allure/jenkins/exceptions/AllureUploadException.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Created by lawyard on 07.10.16.
+ *
+ *  Common exception for allure uploaders
+ *  Use it when catching exceptions different from IOException and InterruptedException
+ */
+public class AllureUploadException extends Exception {
+
+    public AllureUploadException() {}
```
Comment (LawYard) [inline src/main/java/ru/yandex/qatools/allure/jenkins/exceptions/AllureUploadException.java]: Just empty method

**Comment by LawYard** [comment]
Review please

**Comment by baev** [comment]
@LawYard hi

Just so your know - I don't feel like merge this before #62

**Comment by LawYard** [comment]
@baev 
Hi! Thx for your worrying :)

But we successfully used this PR functional last 1 month and wants to merge it before your great huge PR (https://github.com/jenkinsci/allure-plugin/pull/62). Especially considering that work on your PR is not finished yet.

@mavlyutov @pupssman Do u want to join to our discussion?)

**Comment by baev** [comment]
seems like not gonna be merged. Now report is attached to the build as artifact, so you can use other plugin to upload reports to external system and remove them from old builds

**Pull request closed without merging**